From b23eef48ed484b6156d68bf3cd02ce5e0da68a50 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Andreas=20B=C3=BCrkler?= <andreas.buerkler@enclustra.com>
Date: Tue, 28 Jun 2022 13:10:12 +0200
Subject: [PATCH] gpio: microsemi gpio get base dynamically

---
 drivers/gpio/gpio-microsemi.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/gpio/gpio-microsemi.c b/drivers/gpio/gpio-microsemi.c
index 1024ee69477f..5d01c04f97d2 100644
--- a/drivers/gpio/gpio-microsemi.c
+++ b/drivers/gpio/gpio-microsemi.c
@@ -444,7 +444,7 @@ static int microsemi_mss_gpio_probe(struct platform_device *pdev)
 	mss_gpio->gc.get_direction = microsemi_mss_gpio_get_direction;
 	mss_gpio->gc.get = microsemi_mss_gpio_get_value;
 	mss_gpio->gc.set = microsemi_mss_gpio_set_value;
-	mss_gpio->gc.base = 0;
+	mss_gpio->gc.base = (pdev->id < 0) ? of_alias_get_id(node, "gpio") * 32 : pdev->id * 32;
 	mss_gpio->gc.ngpio = ngpio;
 	mss_gpio->gc.label = dev_name(dev);
 	mss_gpio->gc.parent = dev;
-- 
2.25.1

