From 91e98b00fd121da8d4db424d4b4ac23589893c8a Mon Sep 17 00:00:00 2001
From: Andreas Buerkler <andreas.buerkler@enclustra.com>
Date: Mon, 28 Feb 2022 17:57:51 +0100
Subject: [PATCH] PolarFire SoC I2C driver modification for zero sized
 transfers

---
 drivers/i2c/i2c-microchip.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/drivers/i2c/i2c-microchip.c b/drivers/i2c/i2c-microchip.c
index e2a27a7d19..a83fc16dcd 100644
--- a/drivers/i2c/i2c-microchip.c
+++ b/drivers/i2c/i2c-microchip.c
@@ -380,9 +380,6 @@ static int mpfs_i2c_xfer_msg(struct mpfs_i2c_bus *bus, struct i2c_msg *msg)
 	u8 ctrl;
 	int ret;
 
-	if (!msg->len || !msg->buf)
-		return -EINVAL;
-
 	bus->addr = i2c_8bit_addr_from_msg(msg);
 	bus->msg_len = msg->len;
 	bus->buf = msg->buf;
