From 1a339e938bbc6be39360280fd7239b2458297f8c Mon Sep 17 00:00:00 2001
From: Andreas Buerkler <andreas.buerkler@enclustra.com>
Date: Tue, 18 Oct 2022 10:21:47 +0200
Subject: [PATCH] Use only high memory region

---
 arch/riscv/cpu/generic/dram.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/arch/riscv/cpu/generic/dram.c b/arch/riscv/cpu/generic/dram.c
index ea720900e1..7dfe318cc5 100644
--- a/arch/riscv/cpu/generic/dram.c
+++ b/arch/riscv/cpu/generic/dram.c
@@ -23,7 +23,9 @@ int dram_init_banksize(void)
 
 ulong board_get_usable_ram_top(ulong total_size)
 {
-#if !defined(CONFIG_TARGET_MICROCHIP_ICICLE) && !defined(CONFIG_TARGET_MICROCHIP_SEVKIT)
+#if !defined(CONFIG_TARGET_MICROCHIP_ICICLE) && \
+    !defined(CONFIG_TARGET_MICROCHIP_SEVKIT) && \
+    !defined(CONFIG_TARGET_ENCLUSTRA_MERCURY_MP1)
 	/*
 	 * Ensure that we run from first 4GB so that all
 	 * addresses used by U-Boot are 32bit addresses.
